{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from myutils import rmsk\n",
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Creating bed files for LINE-1 Families "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ADAPTED FROM make_txome.py\n",
    "\n",
    "\"\"\"\n",
    "DEFINE FUNCTIONS \n",
    "\"\"\"\n",
    "\n",
    "\n",
    "def rmsk_to_gtf(rmsk: pd.DataFrame) -> pd.DataFrame:\n",
    "    \"convert RepeatMasker output file to gtf\"\n",
    "\n",
    "    rmsk.rename(\n",
    "        columns={\n",
    "            \"genoName\": \"Chromosome\",\n",
    "            \"genoStart\": \"Start\",\n",
    "            \"genoEnd\": \"End\",\n",
    "            \"strand\": \"Strand\",\n",
    "        },\n",
    "        inplace=True,\n",
    "    )\n",
    "\n",
    "    # gene level\n",
    "    rmsk.Start = rmsk.Start - 1  # make 0 based\n",
    "    rmsk[\"gene_id\"] = rmsk[\"repName\"]\n",
    "    rmsk[\"family_id\"] = rmsk[\"repFamily\"]\n",
    "    rmsk[\"class_id\"] = rmsk[\"repClass\"]\n",
    "    rmsk[\"gene_name\"] = rmsk[\"repName\"] + \":TE\"\n",
    "    rmsk = rmsk[\n",
    "        [\n",
    "            \"Chromosome\",\n",
    "            \"Start\",\n",
    "            \"End\",\n",
    "            \"Strand\",\n",
    "            \"gene_id\",\n",
    "            \"family_id\",\n",
    "            \"class_id\",\n",
    "            \"gene_name\",\n",
    "        ]\n",
    "    ].copy()\n",
    "    rmsk[\"Source\"] = \"RepeatMasker\"\n",
    "    rmsk[\"Feature\"] = \"gene\"\n",
    "    rmsk[\"gene_type\"] = \"retrogene\"\n",
    "\n",
    "    # tx level\n",
    "    rmsktx = rmsk.copy().reset_index(drop=True)\n",
    "    rmsktx[\"Feature\"] = \"transcript\"\n",
    "    rmsktx[\"transcript_id\"] = (\n",
    "        rmsktx[\"gene_id\"] + \"_dup\" + rmsktx.groupby(\"gene_id\").cumcount().astype(str)\n",
    "    )\n",
    "\n",
    "    # exon level\n",
    "    rmskex = rmsktx.copy().reset_index(drop=True)\n",
    "    rmskex[\"Feature\"] = \"exon\"\n",
    "    rmskex[\"exon_id\"] = rmskex.transcript_id\n",
    "    rmskex[\"exon_number\"] = 1\n",
    "\n",
    "    return pd.concat([rmsk, rmsktx, rmskex]).sort_values([\"Chromosome\", \"Start\", \"End\"])\n",
    "\n",
    "\n",
    "def gtf_to_bed(rmsk_file, genes, family):\n",
    "    \"convert gtf to bed\"\n",
    "\n",
    "    bed = rmsk_file[rmsk_file[\"repName\"] == family]\n",
    "\n",
    "    bed = bed[[\"genoName\", \"genoStart\", \"genoEnd\", \"repName\", \"strand\"]].copy()\n",
    "    bed.to_csv(f\"../data/{family}.bed\", sep=\"\\t\", index=False, header=False)\n",
    "    return bed\n",
    "\n",
    "    # get rid of chr in chromosome names\n",
    "\n",
    "    truncated_l1 = rmsk_to_gtf(bed[bed[\"is_full_length\"] == False])\n",
    "    full_l1 = rmsk_to_gtf(bed[bed[\"is_full_length\"] == True])\n",
    "\n",
    "    ### Full length intronic and intergenic regions\n",
    "    full_l1[\"Length_TE\"] = full_l1[\"End\"] - full_l1[\"Start\"]\n",
    "\n",
    "    full_l1 = (\n",
    "        pr.PyRanges(full_l1, int64=True)\n",
    "        .join(\n",
    "            pr.PyRanges(genes, int64=True),\n",
    "            how=\"left\",\n",
    "            report_overlap=True,\n",
    "            suffix=\"_gene\",\n",
    "        )\n",
    "        .df\n",
    "    )\n",
    "\n",
    "    full_l1[(full_l1[\"Overlap\"] == full_l1[\"Length_TE\"])].to_csv(\n",
    "        f\"../data/full_intronic_{family}.bed\", sep=\"\\t\", index=False, header=False\n",
    "    )\n",
    "    full_l1[(full_l1[\"Overlap\"] != full_l1[\"Length_TE\"])].to_csv(\n",
    "        f\"../data/full_intergenic_{family}.bed\", sep=\"\\t\", index=False, header=False\n",
    "    )\n",
    "\n",
    "    ### Truncated intronic and intergenic regions\n",
    "    truncated_l1[\"Length_TE\"] = truncated_l1[\"End\"] - truncated_l1[\"Start\"]\n",
    "    truncated_l1 = (\n",
    "        pr.PyRanges(truncated_l1, int64=True)\n",
    "        .join(\n",
    "            pr.PyRanges(genes, int64=True),\n",
    "            how=\"left\",\n",
    "            report_overlap=True,\n",
    "            suffix=\"_gene\",\n",
    "        )\n",
    "        .df\n",
    "    )\n",
    "\n",
    "    truncated_l1[(truncated_l1[\"Overlap\"] == truncated_l1[\"Length_TE\"])].to_csv(\n",
    "        f\"../data/truncate_intronic_{family}.bed\", sep=\"\\t\", index=False, header=False\n",
    "    )\n",
    "\n",
    "    truncated_l1[(truncated_l1[\"Overlap\"] != truncated_l1[\"Length_TE\"])].to_csv(\n",
    "        f\"../data/truncate_intergenic_{family}.bed\", sep=\"\\t\", index=False, header=False\n",
    "    )\n",
    "\n",
    "    return bed"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in gene functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Length_Gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>X</td>\n",
       "      <td>100627108</td>\n",
       "      <td>100639991</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000000003.14</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>12883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>X</td>\n",
       "      <td>100584801</td>\n",
       "      <td>100599885</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000005.5</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>15084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20</td>\n",
       "      <td>50934866</td>\n",
       "      <td>50958555</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000000419.12</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>23689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>169849630</td>\n",
       "      <td>169894267</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000000457.13</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>44637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>169662006</td>\n",
       "      <td>169854080</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000460.16</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>192074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58214</th>\n",
       "      <td>1</td>\n",
       "      <td>157203603</td>\n",
       "      <td>157205062</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000284592.1</td>\n",
       "      <td>RP11-71G12.2</td>\n",
       "      <td>1459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58215</th>\n",
       "      <td>11</td>\n",
       "      <td>1880044</td>\n",
       "      <td>1880147</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000284594.1</td>\n",
       "      <td>MIR7847</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58216</th>\n",
       "      <td>17</td>\n",
       "      <td>75498547</td>\n",
       "      <td>75498628</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000284595.1</td>\n",
       "      <td>MIR6785</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58217</th>\n",
       "      <td>7</td>\n",
       "      <td>102471468</td>\n",
       "      <td>102471531</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000284596.1</td>\n",
       "      <td>MIR4467</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58218</th>\n",
       "      <td>2</td>\n",
       "      <td>1795524</td>\n",
       "      <td>1811526</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000284600.1</td>\n",
       "      <td>RP11-755H23.1</td>\n",
       "      <td>16002</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58219 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome      Start        End Strand             gene_id  \\\n",
       "0              X  100627108  100639991      -  ENSG00000000003.14   \n",
       "1              X  100584801  100599885      +   ENSG00000000005.5   \n",
       "2             20   50934866   50958555      -  ENSG00000000419.12   \n",
       "3              1  169849630  169894267      -  ENSG00000000457.13   \n",
       "4              1  169662006  169854080      +  ENSG00000000460.16   \n",
       "...          ...        ...        ...    ...                 ...   \n",
       "58214          1  157203603  157205062      -   ENSG00000284592.1   \n",
       "58215         11    1880044    1880147      +   ENSG00000284594.1   \n",
       "58216         17   75498547   75498628      +   ENSG00000284595.1   \n",
       "58217          7  102471468  102471531      +   ENSG00000284596.1   \n",
       "58218          2    1795524    1811526      +   ENSG00000284600.1   \n",
       "\n",
       "           gene_name  Length_Gene  \n",
       "0             TSPAN6        12883  \n",
       "1               TNMD        15084  \n",
       "2               DPM1        23689  \n",
       "3              SCYL3        44637  \n",
       "4           C1orf112       192074  \n",
       "...              ...          ...  \n",
       "58214   RP11-71G12.2         1459  \n",
       "58215        MIR7847          103  \n",
       "58216        MIR6785           81  \n",
       "58217        MIR4467           63  \n",
       "58218  RP11-755H23.1        16002  \n",
       "\n",
       "[58219 rows x 7 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf = pr.read_gtf(\"../data/gencode.v26.basic.annotation.gtf.gz\").df\n",
    "genes = []\n",
    "for gene_id, df in gtf.groupby(\"gene_id\"):\n",
    "    assert (\n",
    "        df[\"Strand\"].unique().shape[0] == 1\n",
    "    ), f\"This gene: {gene_id} is on two strands\"\n",
    "\n",
    "    row = {\n",
    "        \"Chromosome\": df[\"Chromosome\"].unique()[0],\n",
    "        \"Start\": min(df[\"Start\"]),\n",
    "        \"End\": max(df[\"End\"]),\n",
    "        \"Strand\": df[\"Strand\"].unique()[0],\n",
    "        \"gene_id\": gene_id,\n",
    "        \"gene_name\": df[\"gene_name\"].unique()[0],\n",
    "    }\n",
    "    genes.append(row)\n",
    "\n",
    "# additional filtering for future methods\n",
    "genes = pd.DataFrame(genes)\n",
    "genes[\"Length_Gene\"] = genes[\"End\"] - genes[\"Start\"]\n",
    "genes[\"Chromosome\"] = genes[\"Chromosome\"].str.replace(\n",
    "    \"chr\", \"\"\n",
    ")  # get rid of chr in chromosome names\n",
    "\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes[\"Chromosome\"] = genes[\"Chromosome\"].str.replace(\"chr\", \"\")\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>swScore</th>\n",
       "      <th>milliDiv</th>\n",
       "      <th>milliDel</th>\n",
       "      <th>milliIns</th>\n",
       "      <th>genoName</th>\n",
       "      <th>genoStart</th>\n",
       "      <th>genoEnd</th>\n",
       "      <th>genoLeft</th>\n",
       "      <th>strand</th>\n",
       "      <th>repName</th>\n",
       "      <th>repStart</th>\n",
       "      <th>repEnd</th>\n",
       "      <th>repLeft</th>\n",
       "      <th>id</th>\n",
       "      <th>repClass</th>\n",
       "      <th>repFamily</th>\n",
       "      <th>length</th>\n",
       "      <th>age</th>\n",
       "      <th>has_promoter</th>\n",
       "      <th>is_full_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>463</td>\n",
       "      <td>13.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>1</td>\n",
       "      <td>10001</td>\n",
       "      <td>10468</td>\n",
       "      <td>248945954</td>\n",
       "      <td>+</td>\n",
       "      <td>(TAACCC)n</td>\n",
       "      <td>1</td>\n",
       "      <td>471</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>None</td>\n",
       "      <td>470</td>\n",
       "      <td>2.980451</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3612</td>\n",
       "      <td>114.0</td>\n",
       "      <td>215.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1</td>\n",
       "      <td>10469</td>\n",
       "      <td>11447</td>\n",
       "      <td>248944975</td>\n",
       "      <td>-</td>\n",
       "      <td>TAR1</td>\n",
       "      <td>399</td>\n",
       "      <td>1712</td>\n",
       "      <td>483</td>\n",
       "      <td>2</td>\n",
       "      <td>Satellite</td>\n",
       "      <td>telo</td>\n",
       "      <td>1229</td>\n",
       "      <td>28.103632</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>484</td>\n",
       "      <td>251.0</td>\n",
       "      <td>132.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>11505</td>\n",
       "      <td>11675</td>\n",
       "      <td>248944747</td>\n",
       "      <td>-</td>\n",
       "      <td>L1MC5a</td>\n",
       "      <td>2382</td>\n",
       "      <td>395</td>\n",
       "      <td>199</td>\n",
       "      <td>3</td>\n",
       "      <td>LINE</td>\n",
       "      <td>L1</td>\n",
       "      <td>196</td>\n",
       "      <td>69.454621</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>239</td>\n",
       "      <td>294.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1</td>\n",
       "      <td>11678</td>\n",
       "      <td>11780</td>\n",
       "      <td>248944642</td>\n",
       "      <td>-</td>\n",
       "      <td>MER5B</td>\n",
       "      <td>74</td>\n",
       "      <td>104</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>DNA</td>\n",
       "      <td>hAT-Charlie</td>\n",
       "      <td>103</td>\n",
       "      <td>84.814840</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>318</td>\n",
       "      <td>230.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>15265</td>\n",
       "      <td>15355</td>\n",
       "      <td>248941067</td>\n",
       "      <td>-</td>\n",
       "      <td>MIR3</td>\n",
       "      <td>119</td>\n",
       "      <td>143</td>\n",
       "      <td>49</td>\n",
       "      <td>5</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>94</td>\n",
       "      <td>62.428022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5520110</th>\n",
       "      <td>256</td>\n",
       "      <td>304.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>57214797</td>\n",
       "      <td>57214899</td>\n",
       "      <td>12516</td>\n",
       "      <td>+</td>\n",
       "      <td>MER5B</td>\n",
       "      <td>1</td>\n",
       "      <td>104</td>\n",
       "      <td>74</td>\n",
       "      <td>4607723</td>\n",
       "      <td>DNA</td>\n",
       "      <td>hAT-Charlie</td>\n",
       "      <td>103</td>\n",
       "      <td>88.594475</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5520111</th>\n",
       "      <td>497</td>\n",
       "      <td>204.0</td>\n",
       "      <td>152.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>57214902</td>\n",
       "      <td>57215093</td>\n",
       "      <td>12322</td>\n",
       "      <td>+</td>\n",
       "      <td>L1MC5a</td>\n",
       "      <td>199</td>\n",
       "      <td>395</td>\n",
       "      <td>2382</td>\n",
       "      <td>4607724</td>\n",
       "      <td>LINE</td>\n",
       "      <td>L1</td>\n",
       "      <td>196</td>\n",
       "      <td>54.111517</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5520112</th>\n",
       "      <td>6064</td>\n",
       "      <td>115.0</td>\n",
       "      <td>204.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>57215117</td>\n",
       "      <td>57216403</td>\n",
       "      <td>11012</td>\n",
       "      <td>+</td>\n",
       "      <td>TAR1</td>\n",
       "      <td>469</td>\n",
       "      <td>2068</td>\n",
       "      <td>43</td>\n",
       "      <td>4607725</td>\n",
       "      <td>Satellite</td>\n",
       "      <td>telo</td>\n",
       "      <td>1599</td>\n",
       "      <td>28.371854</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5520113</th>\n",
       "      <td>744</td>\n",
       "      <td>38.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>57216407</td>\n",
       "      <td>57217320</td>\n",
       "      <td>10095</td>\n",
       "      <td>+</td>\n",
       "      <td>(TTAGGG)n</td>\n",
       "      <td>1</td>\n",
       "      <td>924</td>\n",
       "      <td>0</td>\n",
       "      <td>4607726</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>None</td>\n",
       "      <td>923</td>\n",
       "      <td>8.862834</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5520114</th>\n",
       "      <td>55</td>\n",
       "      <td>105.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>57217321</td>\n",
       "      <td>57217413</td>\n",
       "      <td>10002</td>\n",
       "      <td>+</td>\n",
       "      <td>(GT)n</td>\n",
       "      <td>1</td>\n",
       "      <td>97</td>\n",
       "      <td>0</td>\n",
       "      <td>4607727</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>None</td>\n",
       "      <td>96</td>\n",
       "      <td>25.708447</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5317386 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         swScore  milliDiv  milliDel  milliIns genoName  genoStart   genoEnd  \\\n",
       "0            463      13.0       6.0      17.0        1      10001     10468   \n",
       "1           3612     114.0     215.0      13.0        1      10469     11447   \n",
       "2            484     251.0     132.0       0.0        1      11505     11675   \n",
       "3            239     294.0      19.0      10.0        1      11678     11780   \n",
       "4            318     230.0      37.0       0.0        1      15265     15355   \n",
       "...          ...       ...       ...       ...      ...        ...       ...   \n",
       "5520110      256     304.0      19.0      10.0        Y   57214797  57214899   \n",
       "5520111      497     204.0     152.0      36.0        Y   57214902  57215093   \n",
       "5520112     6064     115.0     204.0      10.0        Y   57215117  57216403   \n",
       "5520113      744      38.0      11.0      25.0        Y   57216407  57217320   \n",
       "5520114       55     105.0      43.0       0.0        Y   57217321  57217413   \n",
       "\n",
       "          genoLeft strand    repName  repStart  repEnd  repLeft       id  \\\n",
       "0        248945954      +  (TAACCC)n         1     471        0        1   \n",
       "1        248944975      -       TAR1       399    1712      483        2   \n",
       "2        248944747      -     L1MC5a      2382     395      199        3   \n",
       "3        248944642      -      MER5B        74     104        1        4   \n",
       "4        248941067      -       MIR3       119     143       49        5   \n",
       "...            ...    ...        ...       ...     ...      ...      ...   \n",
       "5520110      12516      +      MER5B         1     104       74  4607723   \n",
       "5520111      12322      +     L1MC5a       199     395     2382  4607724   \n",
       "5520112      11012      +       TAR1       469    2068       43  4607725   \n",
       "5520113      10095      +  (TTAGGG)n         1     924        0  4607726   \n",
       "5520114      10002      +      (GT)n         1      97        0  4607727   \n",
       "\n",
       "              repClass    repFamily  length        age has_promoter  \\\n",
       "0        Simple_repeat         None     470   2.980451          NaN   \n",
       "1            Satellite         telo    1229  28.103632          NaN   \n",
       "2                 LINE           L1     196  69.454621        False   \n",
       "3                  DNA  hAT-Charlie     103  84.814840          NaN   \n",
       "4                 SINE          MIR      94  62.428022          NaN   \n",
       "...                ...          ...     ...        ...          ...   \n",
       "5520110            DNA  hAT-Charlie     103  88.594475          NaN   \n",
       "5520111           LINE           L1     196  54.111517        False   \n",
       "5520112      Satellite         telo    1599  28.371854          NaN   \n",
       "5520113  Simple_repeat         None     923   8.862834          NaN   \n",
       "5520114  Simple_repeat         None      96  25.708447          NaN   \n",
       "\n",
       "        is_full_length  \n",
       "0                  NaN  \n",
       "1                  NaN  \n",
       "2                False  \n",
       "3                  NaN  \n",
       "4                  NaN  \n",
       "...                ...  \n",
       "5520110            NaN  \n",
       "5520111          False  \n",
       "5520112            NaN  \n",
       "5520113            NaN  \n",
       "5520114            NaN  \n",
       "\n",
       "[5317386 rows x 20 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rmsk_file = rmsk.read_rmsk(\"../data/hg38.fa.out.gz\")\n",
    "rmsk_file[\"genoName\"].unique()\n",
    "\n",
    "rmsk_file[\"genoName\"] = rmsk_file[\"genoName\"].str.replace(\"chr\", \"\")\n",
    "rmsk_file = rmsk_file[rmsk_file[\"genoName\"].isin(genes[\"Chromosome\"].unique())]\n",
    "# keep only standard chromosomes\n",
    "\n",
    "\n",
    "rmsk_file"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make bed files for LINE-1 families"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "L1_families = [\"L1HS\", \"L1PA2\", \"L1PA3\", \"L1PA6\"]\n",
    "\n",
    "for family in L1_families:\n",
    "    df = gtf_to_bed(rmsk_file, genes, family)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genoName</th>\n",
       "      <th>genoStart</th>\n",
       "      <th>genoEnd</th>\n",
       "      <th>repName</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24292</th>\n",
       "      <td>1</td>\n",
       "      <td>12816084</td>\n",
       "      <td>12816355</td>\n",
       "      <td>L1HS</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27163</th>\n",
       "      <td>1</td>\n",
       "      <td>14257711</td>\n",
       "      <td>14257955</td>\n",
       "      <td>L1HS</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65478</th>\n",
       "      <td>1</td>\n",
       "      <td>28985834</td>\n",
       "      <td>28986182</td>\n",
       "      <td>L1HS</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68952</th>\n",
       "      <td>1</td>\n",
       "      <td>30567814</td>\n",
       "      <td>30568750</td>\n",
       "      <td>L1HS</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78927</th>\n",
       "      <td>1</td>\n",
       "      <td>34566056</td>\n",
       "      <td>34572105</td>\n",
       "      <td>L1HS</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5510639</th>\n",
       "      <td>Y</td>\n",
       "      <td>20690314</td>\n",
       "      <td>20690858</td>\n",
       "      <td>L1HS</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5511004</th>\n",
       "      <td>Y</td>\n",
       "      <td>20958692</td>\n",
       "      <td>20959548</td>\n",
       "      <td>L1HS</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5515241</th>\n",
       "      <td>Y</td>\n",
       "      <td>23956872</td>\n",
       "      <td>23959834</td>\n",
       "      <td>L1HS</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5517909</th>\n",
       "      <td>Y</td>\n",
       "      <td>25710292</td>\n",
       "      <td>25713238</td>\n",
       "      <td>L1HS</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5517911</th>\n",
       "      <td>Y</td>\n",
       "      <td>25713299</td>\n",
       "      <td>25713316</td>\n",
       "      <td>L1HS</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1620 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        genoName  genoStart   genoEnd repName strand\n",
       "24292          1   12816084  12816355    L1HS      -\n",
       "27163          1   14257711  14257955    L1HS      +\n",
       "65478          1   28985834  28986182    L1HS      -\n",
       "68952          1   30567814  30568750    L1HS      +\n",
       "78927          1   34566056  34572105    L1HS      -\n",
       "...          ...        ...       ...     ...    ...\n",
       "5510639        Y   20690314  20690858    L1HS      -\n",
       "5511004        Y   20958692  20959548    L1HS      -\n",
       "5515241        Y   23956872  23959834    L1HS      -\n",
       "5517909        Y   25710292  25713238    L1HS      +\n",
       "5517911        Y   25713299  25713316    L1HS      +\n",
       "\n",
       "[1620 rows x 5 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "new_bulk",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
